File: 0021_Merge_Two_Sorted_Lists/Merge_Two_Sorted_Lists.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def merge_two_lists(l1, l2)
  return l2 if l1.nil?
  return l1 if l2.nil?
  result_node = ListNode.new(0)
  current_node = result_node
  last_node = result_node
  while l1 && l2
    if l1.val <= l2.val
      current_node.val = l1.val
      current_node.next = ListNode.new(0)
      l1 = l1.next
    else
      current_node.val = l2.val
      current_node.next = ListNode.new(0)
      l2 = l2.next
    end
    last_node = current_node
    current_node = current_node.next
  end
  while l1
    current_node.val = l1.val
    current_node.next = ListNode.new(0)
    l1 = l1.next
    last_node = current_node
    current_node = current_node.next
  end
  while l2
    current_node.val = l2.val
    current_node.next = ListNode.new(0)
    l2 = l2.next
    last_node = current_node
    current_node = current_node.next
  end
  last_node.next = nil if current_node.val.zero?
  result_node
end

File: 0021_Merge_Two_Sorted_Lists/Merge_Two_Sorted_Lists.py
# -*- coding: utf-8 -*-

"""
Merge two sorted linked lists and return it as a new list. The new list should be made by splicing together the nodes of the first two lists.

Example:

Input: 1->2->4, 1->3->4

Output: 1->1->2->3->4->4
"""


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        result = ListNode(0)
        current = result
        while l1 and l2:
            if l1.val <= l2.val:
                current.next = ListNode(l1.val)
                current = current.next
                l1 = l1.next
            else:
                current.next = ListNode(l2.val)
                current = current.next
                l2 = l2.next
        current.next = l1 or l2 or None
        return result.next
